src/target/riscv/riscv.c

```diff
@@ -2697,7 +2768,158 @@ COMMAND_HANDLER(handle_repeat_read)
 	return result;
 }
 
+COMMAND_HANDLER(handle_memory_sample_command)
+{
+	struct target *target = get_current_target(CMD_CTX);
+
+	if (CMD_ARGC == 0) {
+		for (unsigned i = 0; i < DIM(sample_config.bucket); i++) {
+			if (sample_config.bucket[i].enabled) {
+				command_print(CMD, "bucket %d; address=0x%" TARGET_PRIxADDR "; size=%d", i,
+							  sample_config.bucket[i].address,
+							  sample_config.bucket[i].size_bytes);
+			} else {
+				command_print(CMD, "bucket %d; disabled", i);
+			}
+		}
+		return ERROR_OK;
+	}
+
+	if (CMD_ARGC < 2) {
+		LOG_ERROR("Command requires at least bucket and address arguments.");
+		return ERROR_FAIL;
+	}
+
+	if (riscv_rtos_enabled(target)) {
+		LOG_ERROR("Memory sampling is not supported with `-rtos riscv`.");
+		return ERROR_FAIL;
+	}
+
+	uint32_t bucket;
+	COMMAND_PARSE_NUMBER(u32, CMD_ARGV[0], bucket);
+	if (bucket > DIM(sample_config.bucket)) {
+		LOG_ERROR("Max bucket number is %d.", (unsigned) DIM(sample_config.bucket));
+		return ERROR_COMMAND_ARGUMENT_INVALID;
+	}
+
+	sample_config.bucket[bucket].enabled = true;
+	COMMAND_PARSE_ADDRESS(CMD_ARGV[1], sample_config.bucket[bucket].address);
+
+	if (CMD_ARGC > 2) {
+		COMMAND_PARSE_NUMBER(u32, CMD_ARGV[2], sample_config.bucket[bucket].size_bytes);
+		if (sample_config.bucket[bucket].size_bytes != 4 &&
+				sample_config.bucket[bucket].size_bytes != 8) {
+			LOG_ERROR("Only 4-byte and 8-byte sizes are supported.");
+			return ERROR_COMMAND_ARGUMENT_INVALID;
+		}
+	} else {
+		sample_config.bucket[bucket].size_bytes = 4;
+	}
+
+	if (!sample_buf.buf) {
+		sample_buf.size = 1024 * 1024;
+		sample_buf.buf = malloc(sample_buf.size);
+	}
+	/* Clear the buffer when the configuration is changed. */
+	sample_buf.used = 0;
+
+	sample_config.enabled = true;
+
+	return ERROR_OK;
+}
+
+uint32_t sample_buf_read32(unsigned offset){
+	return sample_buf.buf[offset] |
+		   (((uint32_t)sample_buf.buf[offset + 1]) << 8) |
+		   (((uint32_t)sample_buf.buf[offset + 2]) << 16) |
+		   (((uint32_t)sample_buf.buf[offset + 3]) << 24);
+}
+
+uint64_t sample_buf_read64(unsigned offset){
```

Comment:
It appears `buf_get_u64()` and `buf_get_u32()` could be used here.

Reply:
Done.